components/dashboard/src/settings/Teams.tsx

```diff
@@ -0,0 +1,856 @@
+            if (i >= 0) {
+                next[i] = update(next[i]);
+            }
+            return next;
+        })
+    }
+
+    const slotSetErrorMessage = (slot: TeamSubscriptionSlotResolved) => (err: any) => {
+        updateSlot(slot, (s) => {
+            const result = { ...s }
+            result.errorMsg = err && ((err.data && err.data.msg) || err.message || String(err));
+            result.loading = false;
+            return result;
+        });
+    };
+
+    const slotInputHandler = {
+        assign: (slot: TeamSubscriptionSlotResolved, assigneeIdentifier: string) => {
+            updateSlot(slot, (s) => ({ ...s, loading: true }));
+            getGitpodService().server.tsAssignSlot(slot.teamSubscription.id, slot.id, assigneeIdentifier)
+                .then(slotDoPoll(slot))
+                .catch(slotSetErrorMessage(slot));
+        },
+        reassign: (slot: TeamSubscriptionSlotResolved, newAssigneeIdentifier: string) => {
+            updateSlot(slot, (s) => ({ ...s, loading: true }));
+            getGitpodService().server.tsReassignSlot(slot.teamSubscription.id, slot.id, newAssigneeIdentifier)
+                .then(slotDoPoll(slot))
+                .catch(slotSetErrorMessage(slot));
+        },
+        deactivate: (slot: TeamSubscriptionSlotResolved) => {
+            updateSlot(slot, (s) => ({ ...s, loading: true }));
+            getGitpodService().server.tsDeactivateSlot(slot.teamSubscription.id, slot.id)
+                .then(slotDoPoll(slot))
+                .catch(slotSetErrorMessage(slot));
+        },
+        reactivate: (slot: TeamSubscriptionSlotResolved) => {
+            updateSlot(slot, (s) => ({ ...s, loading: true }));
+            getGitpodService().server.tsReactivateSlot(slot.teamSubscription.id, slot.id)
+                .then(slotDoPoll(slot))
+                .catch(slotSetErrorMessage(slot));
+        }
+    };
+
+    const inputHandler = (ts: TeamSubscription | undefined) => {
+        return {
+            buySlots: async (plan: Plan, quantity: number) => {
+                if (ts) {
+                    // Buy new slots for existing subscription
+                    if (ts.planId !== plan.chargebeeId) {
+                        console.log("Plan IDs do not match!");
+                        return;
+                    }
+
+                    setPendingSlotsPurchase({ ts });
+
+                    getGitpodService().server.tsAddSlots(ts.id, quantity)
+                        .then(() => pollForAdditionalSlotsBought(ts))
+                        .catch((err) => {
+                            setPendingSlotsPurchase(undefined);
+
+                            if (err.code === ErrorCodes.PAYMENT_ERROR) {
+                                alert(`Payment error: ${err.message}`);
+                            }
+                        });
+                } else {
+                    // Buy new subscription + initial slots
+                    setPendingPlanPurchase({ plan });
```

Comment:
Suggestion: You could make any `pendingPlanPurchase` automatically expire after N minutes by saving a timestamp along with the plan:

```suggestion
                    setPendingPlanPurchase({ plan, pendingSince: Date.now() });
```